sdk/typescript/telemetry/telemetry.ts

```diff
@@ -0,0 +1,222 @@
+}
+
+/**
+ * A wrapper around the OpenTelemetry SDK to configure it for Dagger.
+ */
+class DaggerOtelConfigurator {
+  private is_configured = false
+
+  private is_enabled = otelEnabled()
+
+  private sdk?: NodeSDK
+
+  /**
+   * Initialize the Open Telemetry SDK if enabled or not already configured.
+   */
+  initialize() {
+    if (this.is_configured || !this.is_enabled) {
+      return
+    }
+
+    console.debug("Initialiazing telemetry")
+    this.configure()
+
+    // Create the node SDK with the context manager, the resource and the exporter.
+    const sdk = new NodeSDK({
+      resource: new Resource({
+        [SEMRESATTRS_SERVICE_NAME]: "dagger-typescript-sdk",
+      }),
+    })
+
+    // Register the SDK to the OpenTelemetry API if OTEL is enabled
+    sdk.start()
+    console.debug("Telemetry initialized")
+    this.is_configured = true
+    this.sdk = sdk
+  }
+
+  /**
+   * Shutdown the Open Telemetry SDK to flush traces and metrics and close the connection.
+   */
+  async close() {
+    if (this.sdk) {
+      this.sdk.shutdown()
+    }
+  }
+
+  configure() {
+    if (!otelConfigured()) {
+      console.debug("Telemetry not configured")
+      return
+    }
+
+    if (!otelEnabled()) {
+      console.debug("Telemetry disabled")
+      return
+    }
+
+    this.setupEnv()
+  }
+
+  /**
+   * Setup environment for auto-configuring the SDK.
+   */
+  setupEnv() {
+    // Setup a default value for the service name.
+    if (!process.env[OtelEnv.OTEL_SERVICE_NAME]) {
+      process.env[OtelEnv.OTEL_SERVICE_NAME] = SERVICE_NAME
+    }
+
+    // TODO: The insecure flag should be set by the shim instead of the SDK.
+    const endpoints = [
+      OtelEnv.OTEL_EXPORTER_OTLP_ENDPOINT,
+      OtelEnv.OTEL_EXPORTER_OTLP_METRICS_ENDPOINT,
+      OtelEnv.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT,
+      OtelEnv.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT,
+    ]
+
+    endpoints.forEach((endpoint) => {
+      const env = process.env[endpoint]
+      if (env && env.startsWith("http://")) {
+        process.env[OtelEnv.OTEL_EXPORTER_OTLP_INSECURE] = "true"
+      }
+    })
+  }
+}
+
+const configurator = new DaggerOtelConfigurator()
+
+export function initiliaze() {
+  configurator.initialize()
+}
+
+export async function close() {
+  await configurator.close()
+}
+
+/**
+ * Return a tracer to use with Dagger.
+ *
+ * The tracer is automatically initialized if not already done.
+ * As a conveniance function, you can use `withTracingSpan` that automatically close
+ * the span at the end of the function.
+ */
+export function getTracer(): opentelemetry.Tracer {
+  initiliaze()
+
+  const tracer = opentelemetry.trace
+    .getTracerProvider()
+    .getTracer("dagger-typescript-sdk")
+
+  return tracer
+}
+
+export function initContext() {
+  const ctx = opentelemetry.context.active()
+
+  const parentID = process.env.TRACEPARENT
+  if (parentID) {
+    return opentelemetry.propagation.extract(ctx, {
+      traceparent: parentID,
+    })
+  }
+
+  return ctx
+}
```

Comment:
The previous implementation of this was more correct since you kept the way it's used. In Python this has changed as it's no longer returning a context, it's just adding $TRACEPARENT to the global context if it's set, but only the current context isn't valid. That check is missing here too. Python is doing context initialization, but here you're "getting" a context, so it should be `getContext`.

Reply:
I renamed and readd the check